packages/components/src/button/test/index.js

```diff
@@ -12,172 +13,197 @@ import { plusCircle } from '@wordpress/icons';
+				.firstChild;
 
-			expect( button.hasClass( 'gutenberg' ) ).toBe( true );
+			expect( button ).toHaveClass( 'gutenberg' );
 		} );
 
-		it( 'should render and additional WordPress prop of value awesome', () => {
-			const button = shallow( <Button WordPress="awesome" /> ).find(
-				'button'
-			);
+		it( 'should render an additional WordPress prop of value awesome', () => {
+			const button = render( <Button WordPress="awesome" /> ).container
+				.firstChild;
 
-			expect( button.prop( 'WordPress' ) ).toBe( 'awesome' );
+			expect( console ).toHaveErrored();
+			expect( button ).toHaveAttribute( 'wordpress', 'awesome' );
 		} );
 
 		it( 'should render an icon button', () => {
-			const iconButton = shallow( <Button icon={ plusCircle } /> ).find(
-				'button'
-			);
-			expect( iconButton.hasClass( 'has-icon' ) ).toBe( true );
-			expect( iconButton.prop( 'aria-label' ) ).toBeUndefined();
+			const button = render( <Button icon={ plusCircle } /> ).container
+				.firstChild;
+
+			expect( button ).toHaveClass( 'has-icon' );
+			expect( button ).not.toHaveAttribute( 'aria-label' );
 		} );
 
 		it( 'should render a Dashicon component matching the wordpress icon', () => {
-			const iconButton = shallow( <Button icon={ plusCircle } /> ).find(
-				'button'
+			const button = render( <Button icon={ plusCircle } /> ).container
+				.firstChild;
+
+			expect( button ).toContainElement(
+				screen.getByTestId( 'test-icon' )
 			);
-			expect( iconButton.find( 'Icon' ).dive() ).not.toBeNull();
 		} );
 
 		it( 'should render child elements and icon', () => {
-			const iconButton = shallow(
+			const button = render(
 				<Button
 					icon={ plusCircle }
 					children={ <p className="test">Test</p> }
 				/>
-			).find( 'button' );
-			expect( iconButton.find( 'Icon' ).dive() ).not.toBeNull();
-			expect( iconButton.find( '.test' ).shallow().text() ).toBe(
-				'Test'
+			).container.firstChild;
+
+			expect( button ).toContainElement(
+				screen.getByTestId( 'test-icon' )
 			);
+
+			const paragraph = button.childNodes[ 1 ];
+			expect( paragraph ).toHaveClass( 'test' );
+			expect( paragraph ).toHaveTextContent( 'Test' );
 		} );
 
 		it( 'should add an aria-label when the label property is used, with Tooltip wrapper', () => {
-			const iconButton = shallow(
-				<Button icon={ plusCircle } label="WordPress" />
-			).find( 'Tooltip' );
-			expect( iconButton.name() ).toBe( 'Tooltip' );
-			expect( iconButton.prop( 'text' ) ).toBe( 'WordPress' );
-			expect( iconButton.find( 'button' ).prop( 'aria-label' ) ).toBe(
-				'WordPress'
-			);
+			render( <Button icon={ plusCircle } label="WordPress" /> );
+
+			const tooltip = screen.getByTestId( 'test-tooltip' );
+			expect( tooltip ).toBeVisible();
+			expect( tooltip ).toHaveAttribute( 'title', 'WordPress' );
+
+			const button = screen.getByRole( 'button' );
+			expect( tooltip ).toContainElement( button );
+			expect( button ).toHaveAttribute( 'aria-label', 'WordPress' );
```

Comment:
I'm hoping we can manage without mocking Tooltip. For example, this is working for me:

```suggestion
			render( <Button icon={ plusCircle } label="WordPress" /> );
			const button = screen.getByRole( 'button', { name: 'WordPress' } );
			button.focus();
			expect( screen.getByText( 'WordPress' ) ).toBeInTheDocument();
```

Reply:
You definitely have a point here. However, the problem with this type of testing is that we're not sure where `WordPress` appeared. It could have been displayed from the `Tooltip` component from its `title` prop, but it could also have been displayed from another prop. How can we be sure? What you're suggesting simply checks if the text is there, and doesn't care "how" it was displayed. IMHO it also bleeds into the implementation details of a component that we're not interested to test here. Whether `Tooltip` displays its `title` prop somewhere or not, should be subject to its own tests, no? I'm curious to hear what you think about this.

Reply:
Yes, I see your point, too. And together with the mock approach, I find it interesting how it's a true "unit test", in a way that I've never seen in React component testing. (I don't have a lot of Enzyme experience actually!) I guess the weakness here is the unreliability of the mock (the way all mocks are). Realistically, I would want my test to fail if the Tooltip API changed or something and Button stopped showing tooltips. But a mocked test wouldn't fail.

> IMHO it also bleeds into the implementation details of a component that we're not interested to test here. Whether `Tooltip` displays its `title` prop somewhere or not, should be subject to its own tests, no?

It's funny because I would say the mocked test is "testing implementation details" 😆 From the RTL mindset, we want to test that, even in icon button cases where the button doesn't have a visible label, the button is accessible by that label, and a visible label text is in the DOM when the button is focused or hovered. That's it. We don't care whether it even uses the Tooltip component as the implementation. (Not saying that it's inherently unuseful to test that a function calls another function with the correct arguments. It's just not what RTL is designed for. Even in the [FAQ](https://testing-library.com/docs/react-testing-library/faq/) that discusses avoiding mocks, the exception given as an example is a mock meant for bypassing, not for testing arguments.)

Would it be more palatable if we added a `expect( screen.getByText( 'WordPress' ) ).not().toBeInTheDocument();` assertion before the `button.focus()`?

Reply:
I've revisited this and agree with you, so decided to remove the mocks where possible. I've gone that way and now the tests look much better and clearer - especially when we're checking for the tooltip to be hidden first, then after focusing on the button, we're asserting that it's visible.

Reply:
Just going to add my point of view to a super interesting conversation.

I am basically aligned with Lena's point of view here — I usually try to avoid mocking components as much as possible, since the spirit of RTL to feels closer to writing integration tests of the component and its dependencies, rather than an isolated unit test (à la Enzyme's _shallow_ render)

I believe that such tests can be much more valuable and can bring a lot more confidence than a strict unit test. To me, the need for a mock is almost a bit of a "code smell"

Reply:
Thanks for sharing your point of view. I definitely agree with you all here on most points. I disagree on the "code smell" part though - it's not a code smell IMHO. Rather, we're making a trade-off to test as close as the user experiences the component, in favor of stepping out of the testing responsibility of the current component. You are saying mocks are a code smell, others would say that testing other components is a code smell. I say it's a trade-off, and as long as we're making a conscious decision and we've weighed the trade-offs of it, we're good to go.